File: java/components/StrataPersistence/src/strata1/persistence/repository/InputKeeper.java
// ##########################################################################
// # File Name:	.java
// #
// # Copyright:	2011, Sapientia Systems, LLC. All Rights Reserved.
// #
// # License:	This file is part of the StrataEntity Framework.
// #
// #   			The StrataEntity Framework is free software: you 
// #			can redistribute it and/or modify it under the terms of 
// #			the GNU Lesser General Public License as published by
// #    		the Free Software Foundation, either version 3 of the 
// #			License, or (at your option) any later version.
// #
// #    		The StrataEntity Framework is distributed in the 
// #			hope that it will be useful, but WITHOUT ANY WARRANTY; 
// #			without even the implied warranty of MERCHANTABILITY or 
// #			FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser 
// #			General Public License for more details.
// #
// #    		You should have received a copy of the GNU Lesser 
// #			General Public License along with the StrataEntity
// #			Framework. If not, see http://www.gnu.org/licenses/.
// ##########################################################################

package strata1.persistence.repository;

import java.util.Collection;
import java.util.Map;
import java.util.TreeMap;

/**
 * 
 * @author 		
 *     Sapientia Systems 
 * @conventions	
 *     <a href="{@docRoot}/NamingConventions.html">Naming Conventions</a>
 */
public
class InputKeeper
{
	private InputMode           itsMode;
	private Map<String,Object>  itsNamedInputs;
	private Map<Integer,Object> itsPositionalInputs;

	/************************************************************************
	 * Creates a new InputKeeper. 
	 *
	 */
	public 
	InputKeeper()
	{
		super();
		itsMode             = InputMode.NOT_INITIALIZED;
		itsNamedInputs      = new TreeMap<String,Object>();
		itsPositionalInputs = new TreeMap<Integer,Object>();
	}

	/************************************************************************
	 *  
	 *
	 * @param name
	 * @param value
	 * @throws InvalidInputException
	 */
	public void
	setInput(String name,Object value)
		throws InvalidInputException
	{
		checkMode( InputMode.NAMED );
		itsNamedInputs.put( name,value );
	}
	
	/************************************************************************
	 *  
	 *
	 * @param position
	 * @param value
	 * @throws InvalidInputException
	 */
	public void
	setInput(int position,Object value)
		throws InvalidInputException
	{
		checkMode( InputMode.POSITIONAL );
		itsPositionalInputs.put( position,value );
	}
	
	/************************************************************************
	 *  
	 *
	 * @param position
	 * @param value
	 * @throws InvalidInputException
	 */
	public void
	setInput(Collection<Object> inputs)
		throws InvalidInputException
	{
		int i = 0;
		
		checkMode( InputMode.POSITIONAL );
		
		for (Object input: inputs)
			itsPositionalInputs.put( i++,input );
	}
	
	/************************************************************************
	 *  
	 *
	 */
	public void
	clearInputs()
	{
		itsMode = InputMode.NOT_INITIALIZED;
		itsNamedInputs.clear();
		itsPositionalInputs.clear();
	}
	
	/************************************************************************
	 *  
	 *
	 * @return
	 */
	public InputMode
	getMode()
	{
		return itsMode;
	}
	
	/************************************************************************
	 *  
	 *
	 * @return
	 */
	public Map<String,Object>
	getNamedInputs()
	{
		return itsNamedInputs;
	}
	
	/************************************************************************
	 *  
	 *
	 * @return
	 */
	public Collection<Object>
	getPositionalInputs()
	{
		return itsPositionalInputs.values();
	}
	
	/************************************************************************
	 *  
	 *
	 * @param requestedMode
	 * @throws InvalidInputException
	 */
	private void
	checkMode(InputMode requestedMode) throws InvalidInputException
	{
		if ( getMode() == InputMode.NOT_INITIALIZED )
			itsMode = requestedMode;
		else if ( getMode() != requestedMode )
		{
			throw new InvalidInputException( 
				"Cannot mix input modes: current mode is " + 
				getMode() + " requested mode is " + requestedMode + "." );
		}
	}
}


// ##########################################################################

File: legacy/csharp/components/StrataPersistence/Repository/InputKeeper.cs
//  ##########################################################################
//  # File Name: InputKeeper.cs
//  # Copyright: 2011-2018, Sapientia Systems, LLC.
//  ##########################################################################
 
using System;
using System.Collections.Generic;

namespace Strata.Persistence.Repository
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// Helper class that organizes inputs for <c>IFinder{T}</c> 
    /// objects by name or by position.
    /// </summary>
    /// 
    /// <author>JFL</author>
    /// <conventions>$conventionspath$</conventions>
    ///  
    public
    class InputKeeper
    {
	    public  InputMode                  Mode { get; protected set; }
	    private IDictionary<String,Object> namedInputs;
	    private IDictionary<int,Object>    positionalInputs;

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Creates a new InputKeeper. 
        /// </summary>
        /// 
	    public 
	    InputKeeper()
	    {
		    Mode             = InputMode.NOT_INITIALIZED;
		    namedInputs      = new Dictionary<String,Object>();
		    positionalInputs = new Dictionary<int,Object>();
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Stores a named input. 
        /// </summary>
        /// 
        /// <param name="name">input name</param>
        /// <param name="value">input value</param>
        /// 
        /// <exception cref="InvalidInputException">
        /// Mixing input modes (named vs positional) is invalid.
        /// </exception>
        /// 
	    public void
	    SetInput(String name,Object value)
	    {
		    CheckMode( InputMode.NAMED );
		    namedInputs.Add( name,value );
	    }
	
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        ///  Stores a positional input
        /// </summary>
        /// 
        /// <param name="position">input position</param>
        /// <param name="value">input value</param>
        /// 
        /// <exception cref="InvalidInputException">
        /// Mixing input modes (named vs positional) is invalid.
        /// </exception>
        /// 
 	    public void
	    SetInput(int position,Object value)
	    {
		    CheckMode( InputMode.POSITIONAL );
		    positionalInputs.Add( position,value );
	    }
	
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Stores a collection of positional inputs 
        /// </summary>
        /// 
        /// <param name="inputs">positional inputs</param>
        /// 
        /// <exception cref="InvalidInputException">
        /// Mixing input modes (named vs positional) is invalid.
        /// </exception>
        /// 
	    public void
	    SetInput(ICollection<Object> inputs)
	    {
		    int i = 0;
		
            ClearInputs();
		    CheckMode( InputMode.POSITIONAL );
		
		    foreach (Object input in inputs)
			    positionalInputs.Add( i++,input );
	    }
	
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Clears the stored inputs and resets the input mode. 
        /// </summary>
        /// 
	    public void
	    ClearInputs()
	    {
		    Mode = InputMode.NOT_INITIALIZED;
		    namedInputs.Clear();
		    positionalInputs.Clear();
	    }
	
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Gets the stored named inputs. 
        /// </summary>
        /// 
        /// <returns>the named inputs</returns>
        /// 
	    public IDictionary<String,Object>
	    GetNamedInputs()
	    {
		    return namedInputs;
	    }
	
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Gets the stored positional inputs. 
        /// </summary>
        /// 
        /// <returns>the positional inputs</returns>
        /// 
	    public ICollection<Object>
	    GetPositionalInputs()
	    {
		    return positionalInputs.Values;
	    }
	
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Validates the requested input mode. 
        /// </summary>
        /// 
        /// <param name="requestedMode">requested input mode</param>
        /// 
        /// <exception cref="InvalidInputException">
        /// Mixing input modes (named vs positional) is invalid.
        /// </exception>
        /// 
	    private void
	    CheckMode(InputMode requestedMode)
	    {
		    if ( Mode == InputMode.NOT_INITIALIZED )
			    Mode = requestedMode;
		    else if ( Mode != requestedMode )
		    {
			    throw new InvalidInputException( 
				    "Cannot mix input modes: current mode is " + 
				    Mode + " requested mode is " + requestedMode + "." );
		    }
	    }

    }
}

//  ##########################################################################
